File: go projects/numerical methods/root finding/bisection/utility.go
package main
import "math"


func findInterval(function func(float64) float64, window [2]float64) []map[string]float64 {

	var interval []map[string]float64
	initial := window[0]
	length := math.Abs(window[1] - initial)
	step := 1.0
	delta := float64(length) / step
	x := initial

	for (delta > 1) {

		step += 1
		delta = float64(length) / float64(step)

	}

	for (x < window[1]) {

		x += delta

		if (function(initial)*function(x) < 0) {

			interval = append(interval, map[string]float64{ "lower": initial,
															"upper": x })

		}

		initial = x

	}

	return interval

}

func bisection(function func(float64) float64, window [2]float64, precision float64) ([]float64, float64, int) {

	var roots []float64
	var left float64
	var right float64
	var center float64
	var residual float64
	var previous float64
	iteration := 0
	interval := findInterval(function, window)

	for _, bracket := range interval {

		left = bracket["lower"]
		right = bracket["upper"]
		center = (left + right) / 2
		residual = math.Abs(2.0*precision)
		previous = center

		for (residual > precision) {

			if ((function(left)*function(center) == 0) ||
				(function(right)*function(center) == 0)) {

				roots = append(roots, center)
				iteration += 1
				break

			} else if (function(left)*function(center) < 0) {

				right = center

			} else {

				left = center

			}

			center = (left + right) / 2
			residual = math.Abs((center - previous) / center)
			previous = center
			iteration += 1

			if (residual < precision) {

				roots = append(roots, center)

			}

		}

	}

	return roots, residual, iteration

}

File: rust projects/numerical methods/root finding/secant/src/utility.rs
#![allow(non_snake_case)]
#![allow(unused_parens)]
#![allow(clippy::needless_return)]
#![allow(clippy::type_complexity)]
use std::collections::HashMap;


pub fn secant(function: fn(f64) -> f64, window: [f64; 2], precision: f64) -> (Vec<f64>, f64, u64) {

	let mut roots = vec![];
	let mut iterations: u64 = 0;
	let mut residual: f64 = 0.0;
	let intervals = findInterval(function, window);

	for interval in intervals.iter() {

		let mut x0 = *interval.get("lower").unwrap();
		let mut x1 = *interval.get("upper").unwrap();
		residual = 2.0*precision;

		while (residual.abs() > precision) {

			let x = x1 - function(x1)*(x1 - x0) / (function(x1) - function(x0));
			residual = (x - x1) / x;
			iterations += 1;
			x0 = x1;
			x1 = x;

			if (residual.abs() < precision) {

				roots.push(x);

			}

		}

	}

	return (roots, residual, iterations);

}

fn findInterval(function: fn(f64) -> f64, window: [f64; 2]) -> Vec<HashMap<String, f64>> {

	let mut domain: Vec<HashMap<String, f64>> = Vec::new();
	let mut interval: HashMap<String, f64> = HashMap::new();
	let mut initial: f64 = window[0];
	let width: f64 = (window[1] - initial).abs();
	let mut step: f64 = 1.0;
	let mut delta = width / step;
	let mut x: f64 = initial;

	while (delta > 1.0) {

		step += 1.0;
		delta = width / step;
	
	}

	while (x < window[1]) {

		x += delta;

		if (function(initial)*function(x) < 0.0) {

			interval.insert("lower".to_string(), initial);
			interval.insert("upper".to_string(), x);
			domain.push(interval.clone());

		}

		initial = x

	}

	return domain;

}
